# Deep Learning: Charity Funding Predictor

## Contributor: Sam Espe

### Step 3: Optimizing the Model - Attempt 2

For this attempt, I am going to try adding another hidden layer to see if I can increase the accuracy and decrease the loss of my model.

Reading in data and model

In [1]:
# Import dependencies

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

In [2]:
# Read in charity_data.csv


application_df = pd.read_csv('https://charity-funding-data-set.s3.us-east-2.amazonaws.com/charity_data.csv')
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [4]:
# Load in the saved model from the .h5 file

optimization_2_model = tf.keras.models.load_model("/content/gdrive/MyDrive/Data Visualization & Analysis Boot Camp/Deep-Learning-Charity-Funding/AlphabetSoupCharity.h5")

In [5]:
optimization_2_model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 150)               7050      
                                                                 
 dense_1 (Dense)             (None, 150)               22650     
                                                                 
 dense_2 (Dense)             (None, 1)                 151       
                                                                 
Total params: 29,851
Trainable params: 29,851
Non-trainable params: 0
_________________________________________________________________


### Pre-process the Data

Adjusting for model modifications



In [6]:
# Split off IS_SUCCESSFUL (target) from the feature variables
y = application_df["IS_SUCCESSFUL"]
X = application_df.drop(columns= "IS_SUCCESSFUL")

In [7]:
# Remove unnecessary columns (EIN & NAME)
X = X.drop(columns= ["EIN", "NAME"])

In [8]:
# Determine number of unique values for each column
X.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
dtype: int64

For non-numerical columns with more than 10 unique values (Application Type and Classification), bin low-frequency values into an "Other" category.

In [17]:
# Find distribution of unique values of Application Type
X["APPLICATION_TYPE"].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [18]:
# Save the value counts series

app_type = X["APPLICATION_TYPE"].value_counts()

In [19]:
# Turn the series into a dictionary

app_dict = app_type.to_dict()

In [20]:
# If the application type has fewer than 100 instances, add it to the list of low-frequency application types.

small_apps = []

for key, value in app_dict.items():
  if int(value) < 100:
    small_apps.append(key)

In [21]:
# If an entry in the Application Type column of the dataframe contains an application type that is in the low-frequency list, replace the value with "Other".

for line in range(len(X)):
    if X.iloc[line, 0] in small_apps:
        X.iloc[line, 0] = "Other"

In [22]:
# Find the distribution of unique values of Classification

X["CLASSIFICATION"].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [23]:
# Save the value counts of the unique classifications

classes = X["CLASSIFICATION"].value_counts()

In [24]:
# Turn the series into a dictionary

classes_dict = classes.to_dict()

In [25]:
# If the class has fewer than 200 instances, add the class to the list of low-frequency classes.

small_classes = []

for key, value in classes_dict.items():
    if int(value) < 200:
        small_classes.append(key)

In [26]:
# If an entry in the Classification column of the dataframe contains a classification that is in the low-frequency list, replace the value with "Other".

for line in range(len(X)):
    if X.iloc[line, 2] in small_classes:
        X.iloc[line, 2] = "Other"

In [27]:
# One-hot encode categorical variables (Application Type, Classification, Use Case, Organizaton Type, Income Amt, Special Considerations)

X_encoded_df = pd.get_dummies(X)

In [28]:
# The Special Considerations column was encoded twice: once where Y was encoded as 1, and once where N was encoded as 1. I am removing the column where N was encoded as 1 to
# avoid having duplicate data.

X_dummies = X_encoded_df.drop(columns = "SPECIAL_CONSIDERATIONS_N")

In [29]:
# Split data set into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(X_encoded_df, y, random_state = 7)

In [30]:
# Scale the data

scaler = StandardScaler()
scaler.fit(X_train)

StandardScaler()

In [31]:
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

### Compile, Train, and Evaluate the Model

In [10]:
# Create Keras sequential model

optimization_2_model = tf.keras.models.Sequential()

In [11]:
# Create input layer

optimization_2_model.add(tf.keras.layers.Dense(units = 150, activation = "relu", input_dim = 46))

In [12]:
# Create first hidden layer

optimization_2_model.add(tf.keras.layers.Dense(units = 150, activation = "relu"))

In [13]:
# Create second hidden layer

optimization_2_model.add(tf.keras.layers.Dense(units = 100, activation = "relu"))

In [14]:
# Create output layer

optimization_2_model.add(tf.keras.layers.Dense(units = 1, activation = "sigmoid"))

In [15]:
# Check structure of the model

optimization_2_model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 150)               7050      
                                                                 
 dense_1 (Dense)             (None, 150)               22650     
                                                                 
 dense_2 (Dense)             (None, 100)               15100     
                                                                 
 dense_3 (Dense)             (None, 1)                 101       
                                                                 
Total params: 44,901
Trainable params: 44,901
Non-trainable params: 0
_________________________________________________________________


In [33]:
# Compile the model

optimization_2_model.compile(loss = "binary_crossentropy", optimizer = "adam", metrics = ["accuracy"])

In [34]:
# Train the model

fit_model = optimization_2_model.fit(X_train_scaled, y_train, epochs = 100)

Epoch 1/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5682 - accuracy: 0.7206
Epoch 2/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5538 - accuracy: 0.7292
Epoch 3/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5505 - accuracy: 0.7307
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5492 - accuracy: 0.7331
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5479 - accuracy: 0.7329
Epoch 6/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5463 - accuracy: 0.7331
Epoch 7/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5455 - accuracy: 0.7341
Epoch 8/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5453 - accuracy: 0.7333
Epoch 9/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5444 - accuracy: 0.7347
Epoch 10/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5439 - accura

In [41]:
# Evaluate the model using the testing data

model_loss, model_accuracy = optimization_2_model.evaluate(X_test_scaled, y_test, verbose = 2)

print(f'Loss: {round(model_loss * 100, 2)}%, Accuracy: {round(model_accuracy * 100, 2)}%')

268/268 - 0s - loss: 0.6138 - accuracy: 0.7276 - 372ms/epoch - 1ms/step
Loss: 61.38%, Accuracy: 72.76%


In [36]:
# Save and export results as HDF5 file

optimization_2_model.save("Alphabet_Soup_Charity_Optimization_2.h5")

The original model had an accuracy of 72.73% and a loss of 58.55%. This model has an accuracy of 72.76% and a loss of 61.38%. This model did worse than the original model. It is possible that it was somewhat overtrained because the accuracy of the model on the training set was a little better than on the testing set, and the loss on the training set was much better than on the testing set.